enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,173 @@
+
+This enhancement proposal describes the migration of OpenStack platform from the deprecated [in-tree cloud provider](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#openstack) to the [Cloud Controller Manager](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-openstack-cloud-controller-manager.md#get-started-with-external-openstack-cloud-controller-manager-in-kubernetes) service that implements `external cloud provider` interface.
+
+## Motivation
+
+Using Cloud Controller Managers (CCMs) is the Kubernetes' [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with underlying cloud platforms as it provides more flexibility and freedom for developers. It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed in Kubernetes 1.21. But they are still used in OpenShift and we must start a smooth migration towards CCMs. As a pioneer platform, it is proposed to use OpenStack.
+
+### Goals
+
+- OpenShift, when installed on OpenStack, doesn't utilize the deprecated in-tree cloud provider and uses OpenStack CCM instead.
+- The CCM is deployed and managed by the related cluster operator - `cluster-cloud-controller-manager-operator`.
+- There is an upgrading path from older OpenShift versions with the in-tree cloud provider to CCM'ed ones.
+
+### Non-Goals
+
+- Cinder CSI driver integration is out of scope of this work.
+- Deprecating `hyperkube` binary and switching to standalone binaries for all related components (`kube-apiserver`, `kube-scheduler`, `kube-controller-manager` and others) is also not a goal.
+
+## Proposal
+
+Our main goal is to start using Cloud Controller Manager in OpenShift 4 on OpenStack. So we are going to use the implementation available in upstream as a part of [cloud-provider-openstack](https://github.com/kubernetes/cloud-provider-openstack/tree/master/pkg/cloud-controller-manager) repo.
+
+To maintain the lifecycle of the CCM we want to implement a cluster operator, that will handle all administrative tasks: deploy, restore, upgrade, healthchecks, and so on. Later this operator can be reused to manage CCMs for other platforms (AWS, Azure, GCP and so on).
+
+### Action plan
+
+#### Build OpenStack CCM image by OCP automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Test Cloud Controller Manager manually
```

Comment:
Do we expect changes on how the CCM gets its configuration, or will it happily consume the same `cloud.conf` file that was used for the in-tree cloud provider?

Reply:
i think the configuration should be the same as before (for the in-tree cp). Cinder part will be ignored.

Reply:
I explained this part in the proposal